Implement database iterator

jjxtra opened the pull request:
I would appreciate code review on this. It passes all tests.

I would love to see this in a nuget package soon.

As a side note, some of the threading tests are failing. I did not investigate that.

jjxtra commented:
I would suggest a squash and merge on this to avoid soiling the commit history

oschwald commented on MaxMind.Db/CachedDictionary.cs:
```diff
@@ -0,0 +1,423 @@
+        /// <summary>
+        /// The first node in the priority list
+        /// </summary>
+        protected LinkedListNode<KeyValuePair<TKey, TValue>> FirstNode
```
I don't think this is used.

oschwald commented on MaxMind.Db/CachedDictionary.cs:
```diff
@@ -0,0 +1,423 @@
+        /// <summary>
+        /// Current comparer
+        /// </summary>
+        public IEqualityComparer<TKey> Comparer
```
This and the following three are also not used.

oschwald commented on MaxMind.Db/Reader.cs:
```diff
@@ -170,8 +212,75 @@ private void Dispose(bool disposing)
+        /// </summary>
+        /// <param name="cacheSize">The size of the data cache. This can greatly speed enumeration at the cost of memory usage.</param>
+        /// <returns>Enumerator for all data nodes</returns>
+        public IEnumerable<Reader.ReaderIteratorNode<T>> FindAll<T>(int cacheSize = 16384) where T : class
```
A test that used something other than `Dictionary<string, object>` might be good. In particular, one using the `TypeHolder` class with the `MaxMind-DB-test-decoder.mmdb` database.

jjxtra commented on MaxMind.Db/CachedDictionary.cs:
```diff
@@ -0,0 +1,423 @@
+    /// A dictionary that caches up to N values in memory. Once the dictionary reaches N count, the last item in the internal list is removed.
+    /// New items are always added to the start of the internal list.
+    /// </summary>
+    internal class CachedDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IDisposable
```
Something I use in other projects, can tag it with MIT license. All code written by me in my spare time for free.

oschwald replied:
Thanks! If you are the owner of the code, would you be willing to place it under the Apache 2 license to match the other code in the repo and for the benefit of the patent grant?

jjxtra replied:
Sure no problem

jjxtra replied:
It is done.

jjxtra replied:
Commented code eradicated

jjxtra commented:
I've implemented most of these changes. I've moved the code to the benchmark project, but it is commented out for now. Not sure how useful it is at the moment, perhaps someone else can add it later if needed.

oschwald commented:
Looks good. Perhaps the benchmark code could be moved to a separate PR or an issue so that we don't have commented out code lying around in  the codebase. Other than that and the license question, I am happy to merge this.

jjxtra commented:
Ready to merge when you are.

oschwald commented:
Looks great! Thanks. I am squashing it so that the binary database does not end up in the Git history.
